packages/cli/yaml/docs-validator/src/rules/valid-markdown-link/valid-markdown-link.ts

```diff
@@ -1,25 +1,198 @@
 
 export const ValidMarkdownLinks: Rule = {
     name: "valid-markdown-links",
-    create: () => {
+    create: async ({ workspace, fernWorkspaces }) => {
+        const instanceUrls = getInstanceUrls(workspace);
+
+        const url = instanceUrls[0] ?? "http://localhost";
+
+        const ossWorkspaces: OSSWorkspace[] = [];
+
+        const docsDefinitionResolver = new DocsDefinitionResolver(
+            url,
+            workspace,
+            ossWorkspaces,
+            fernWorkspaces,
+            NOOP_CONTEXT
+        );
+
+        const resolvedDocsDefinition = await docsDefinitionResolver.resolve();
+
+        if (!resolvedDocsDefinition.config.root) {
+            throw new Error("Root node not found");
+        }
+
+        // TODO: this is a bit of a hack to get the navigation tree. We should probably just use the navigation tree
+        // from the docs definition resolver, once there's a light way to retrieve it.
+        const root = FernNavigation.migrate.FernNavigationV1ToLatest.create().root(resolvedDocsDefinition.config.root);
+
+        // all the page slugs in the docs:
+        const collector = FernNavigation.NodeCollector.collect(root);
+
+        const visitableSlugs = new Set<string>();
+        const absoluteFilePathsToSlugs = new Map<AbsoluteFilePath, string[]>();
+        const endpoints: FernNavigation.NavigationNodeApiLeaf[] = [];
+        collector.slugMap.forEach((node, slug) => {
+            // NOTE: even if the node is not a page, it can still be "visitable" because it will redirect to another page.
+            visitableSlugs.add(slug);
+
+            if (node == null || !FernNavigation.isPage(node)) {
+                return;
+            }
+
+            if (FernNavigation.isApiLeaf(node)) {
+                endpoints.push(node);
+            }
+
+            const pageId = FernNavigation.getPageId(node);
+            if (pageId == null) {
+                return;
+            }
+
+            const absoluteFilePath = join(workspace.absoluteFilePath, RelativeFilePath.of(pageId));
+            const slugs = absoluteFilePathsToSlugs.get(absoluteFilePath) ?? [];
+            slugs.push(slug);
+            absoluteFilePathsToSlugs.set(absoluteFilePath, slugs);
+        });
+
         return {
             markdownPage: async ({ content, absoluteFilepath }) => {
+                const slugs = absoluteFilePathsToSlugs.get(absoluteFilepath);
+
+                // if this happens, this probably means that the current file is omitted from the docs navigation
+                // most likely due to a slug collision. This should be handled in a different rule.
+                if (!slugs || slugs.length === 0) {
+                    return [];
+                }
+
+                // if this file cannot be indexed (noindex=true, or hidden=true), then we don't need to check for links
+                // since it will not hurt SEO.
+                // TODO: we should at least report this somehow, since it's a bad UX.
```

Comment:
Should we report this as a warning now that we've distinguished `fatal`, `error`, and `warning` more clearly?

Reply:
@abvthecity what do you think about making these ones `severity: "warning"` while keeping the others `severity: "error"` (though not `"fatal"`) in our new paradigm?